src/coins.cpp

```diff
@@ -488,19 +506,59 @@ double CCoinsViewCache::GetPriority(const CTransaction &tx, int nHeight, CAmount
 CCoinsViewCursor::~CCoinsViewCursor() {}
 static const size_t nMaxOutputsPerBlock =
     DEFAULT_LARGEST_TRANSACTION / ::GetSerializeSize(CTxOut(), SER_NETWORK, PROTOCOL_VERSION);
-const Coin &AccessByTxid(const CCoinsViewCache &view, const uint256 &txid)
-{
-    // We may return a reference to a coin cache entry so therefore we must make sure the caller
-    // has taken a lock on cs_utxo.
-    AssertLockHeld(view.cs_utxo);
 
+CoinAccessor::CoinAccessor(const CCoinsViewCache &view, const uint256 &txid) : cache(&view), lock(cache->csCacheInsert)
+{
+    cache->cs_utxo.lock_shared();
+    EnterCritical("CCoinsViewCache.cs_utxo", __FILE__, __LINE__, (void *)(&cache->cs_utxo));
     COutPoint iter(txid, 0);
+    coin = &emptyCoin;
     while (iter.n < nMaxOutputsPerBlock)
     {
-        const Coin &alternate = view.AccessCoin(iter);
+        const Coin &alternate = view._AccessCoin(iter);
         if (!alternate.IsSpent())
-            return alternate;
+        {
+            coin = &alternate;
+            return;
+        }
         ++iter.n;
     }
-    return coinEmpty;
+}
+
+CoinAccessor::CoinAccessor(const CCoinsViewCache &cacheObj, const COutPoint &output)
+    : cache(&cacheObj), lock(cache->csCacheInsert)
+{
+    cache->cs_utxo.lock_shared();
+    EnterCritical("CCoinsViewCache.cs_utxo", __FILE__, __LINE__, (void *)(&cache->cs_utxo));
+    it = cache->FetchCoin(output, &lock);
+    if (it != cache->cacheCoins.end())
+        coin = &it->second.coin;
+    else
+        coin = &emptyCoin;
+}
+
+CoinAccessor::~CoinAccessor()
+{
+    coin = nullptr;
+    LeaveCritical();
+    cache->cs_utxo.unlock_shared();
```

Comment:
In terms of where to put the enter/leave critical...should't the EnterCritical() come before actually locking , and the LeaveCritical(), after the lock released.  That would then be consistent with what is in sync.h in ENTER_CRITICAL_SECTION and LEAVE_CRITICAL_SECTION.  Timing wise that would make sense I think because the lockstack could be showing that the lock was released when it hasn't been yet?

Reply:
OK this is why I had it in the other direction:

Let's look at the original order and imagine 2 threads, A, B executing this at the same time.
EnterCritical() // push and check lock stack
lock()

So the order could be:
A EnterCritical()
B EnterCritical()  # oops code is going to think you entered the crit-sec but really B is waiting for it.
A lock()  # succeeds
B lock()  # blocks

Turns out that this doesn't matter because the code doesn't care that we haven't entered the section but really we are blocked waiting to enter. 

Either statement order will work, because the code does not care about timing.  It only looks at locking order.  But I will reverse the order to keep it consistent, in case someday a change is added which does create an order issue.